geospaas_processing/cli/copy.py

```diff
@@ -0,0 +1,108 @@
+"""
+Download files that are selected from the database using input criteria.
+"""
+import argparse
+import json
+import ntpath
+import os
+from os import path
+import shutil
+
+import django
+import geospaas_processing.cli.download as download_cli
+from django.contrib.gis.geos import GEOSGeometry
+from geospaas.catalog.managers import LOCAL_FILE_SERVICE
+from geospaas.catalog.models import Dataset
+
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'geospaas_processing.settings')
+django.setup()
+
+
+def main():
+    """
+    Copy the files based on the addressed stored in database to the destination folder.
+    The destination folder as well as criteria for finding the desired files are input of this
+    function.
+    """
+    arg = cli_parse_args()
+    cumulative_query = json.loads(arg.query) if arg.query else {}
+    if arg.geometry:
+        cumulative_query['geographic_location__geometry__intersects'] = GEOSGeometry(arg.geometry)
+    designated_begin, designated_end = download_cli.find_designated_time(
+        arg.rel_time_flag, arg.begin, arg.end)
+    cumulative_query['time_coverage_start__gte'] = designated_begin
+    cumulative_query['time_coverage_end__lte'] = designated_end
+    datasets = Dataset.objects.filter(**cumulative_query)
+    for dataset in datasets:
+        try:
+            source_paths = dataset.dataseturi_set.filter(service=LOCAL_FILE_SERVICE)
+        except IndexError:
+            continue
+        for source_path in source_paths:
```

Comment:
No. This way, we are sure that even if only one local address is valid(I mean is placed there and ready to copy), then copy action will be accomplished. If there are some addresses in the database but the file is moved or maybe dabase is being manipulated by other workflows, this way of coding makes us assure that the file is copied.

Otherwise, as soon as a invalid local address is found in the database, the file is not copied.

Fortunately, the `shutil.copy` does not create duplicate of files if we run the code several times. This means we have idempotency.

Reply:
`shutil.copy()` replaces the destination file if it already exists, so it would be best to stop the loop after the first successful copy.

Reply:
I agree. Better to add a `break`

Reply:
Done, I have added one more line based on your suggestion to prevent `shutil.copy` from replacing!